modules/tinymce/src/plugins/imagetools/main/ts/core/Actions.ts

```diff
@@ -20,43 +21,52 @@ import * as ImageSize from './ImageSize';
 
 let count = 0;
 
-const getFigureImg = (elem) => SelectorFind.child(SugarElement.fromDom(elem), 'img');
+const enum FileExtractType {
+  Name = 2,
+  NameExt = 1
+}
 
-const isFigure = (editor: Editor, elem) => editor.dom.is(elem, 'figure');
+const getFigureImg = (elem: HTMLElement) => SelectorFind.child<HTMLImageElement>(SugarElement.fromDom(elem), 'img');
 
-const getEditableImage = (editor: Editor, elem) => {
-  const isImage = (imgNode) => editor.dom.is(imgNode, 'img:not([data-mce-object],[data-mce-placeholder])');
-  const isEditable = (imgNode) => isImage(imgNode) && (isLocalImage(editor, imgNode) || isCorsImage(editor, imgNode) || Settings.getProxyUrl(editor));
+const isFigure = (editor: Editor, elem: Node): elem is HTMLElement => editor.dom.is(elem, 'figure');
 
-  if (isFigure(editor, elem)) {
-    const imgOpt = getFigureImg(elem);
-    return imgOpt.map((img) => isEditable(img.dom) ? Optional.some(img.dom) : Optional.none());
+const isImage = (editor: Editor, imgNode: Node): imgNode is HTMLImageElement => editor.dom.is(imgNode, 'img:not([data-mce-object],[data-mce-placeholder])');
+
+const getEditableImage = (editor: Editor, node: Node): Optional<HTMLImageElement> => {
+  const isEditable = (imgNode: Node): imgNode is HTMLImageElement =>
+    isImage(editor, imgNode) && (isLocalImage(editor, imgNode) || isCorsImage(editor, imgNode) || Type.isNonNullable(Settings.getProxyUrl(editor)));
+
+  if (isFigure(editor, node)) {
+    return getFigureImg(node).bind((img) => {
+      return isEditable(img.dom) ? Optional.some(img.dom) : Optional.none();
+    });
+  } else {
+    return isEditable(node) ? Optional.some(node) : Optional.none();
   }
-  return isEditable(elem) ? Optional.some(elem) : Optional.none();
 };
 
-const displayError = function (editor: Editor, error) {
+const displayError = function (editor: Editor, error: string) {
   editor.notificationManager.open({
     text: error,
     type: 'error'
   });
 };
 
-const getSelectedImage = (editor: Editor): Optional<SugarElement> => {
+const getSelectedImage = (editor: Editor): Optional<SugarElement<HTMLImageElement>> => {
   const elem = editor.selection.getNode();
-  if (isFigure(editor, elem)) {
-    return getFigureImg(elem);
-  } else {
+  const figureElm = editor.dom.getParent(elem, 'figure.image');
+  if (figureElm !== null && isFigure(editor, figureElm)) {
+    return getFigureImg(figureElm);
+  } else if (isImage(editor, elem)) {
     return Optional.some(SugarElement.fromDom(elem));
+  } else {
+    return Optional.none();
   }
 };
 
-const extractFilename = function (editor: Editor, url: string) {
-  const m = url.match(/\/([^\/\?]+)?\.(?:jpeg|jpg|png|gif)(?:\?|$)/i);
-  if (m) {
-    return editor.dom.encode(m[1]);
-  }
-  return null;
+const extractFilename = function (editor: Editor, url: string, group: FileExtractType) {
+  const m = url.match(/(?:\/|^)(([^\/\?]+)\.(?:[a-z0-9.]+))(?:\?|$)/i);
```

Comment:
This fixes one case that we missed when making the `image_file_types` changes, as we need to handle more than just these three formats now.